### PR TITLE
Run an allowed set of e2e tests in parallel to speed up tests

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

We have a lot of trivial tests which don't do much, except for waiting for a small cirros VM to start. This PR changes our ginkgo test suite to be capable of running in parallel. It will run a first set of 245 allowed tests in parallel to speed up the test runs. In order to achieve this, the following was done:

 * Marks all tests with a `[Serial]` tag and explicitly allows 245 reviewed tests for parallel runs.
 * Three ginkgo-execution nodes will be spawned for parallel runs. This means that three tests can be executed in parallel.
 * Every ginkgo-node runs on its own test namespaces (`kubevirt-test-defaultX` and `kubevirt-test-alternativeX`)
 * The kubernetes-dumper-reporter logs where it will log the artifacts if they fail, to make it still possible to locate the matching artifacts
 * KIND lanes are not executed in parallel. They heavily consume sysctl controlled resources, which can lead to issues with increased load
 * Pulls in the ginkgo-source for the ginkgo-binary and builds it as part of `make build-functest`, to make it unnecessary to have the ginkgo binary available in the test environment. This is following the kuberentes example.

On a parallel run, the expected output looks similar to this:

```
$ find _out/artifacts/
_out/artifacts/
_out/artifacts/partial.junit.functest.1.xml
_out/artifacts/partial.junit.functest.3.xml
_out/artifacts/partial.junit.functest.2.xml
_out/artifacts/partial.junit.functest.xml
_out/artifacts/junit.merged.functest.xml
_out/artifacts/k8s-reporter
_out/artifacts/k8s-reporter/2
_out/artifacts/k8s-reporter/2/1_nodes.log
_out/artifacts/k8s-reporter/2/1_pvs.log
_out/artifacts/k8s-reporter/2/1_pvcs.log
_out/artifacts/k8s-reporter/2/1_events.log
_out/artifacts/k8s-reporter/2/1_endpoints.log
_out/artifacts/k8s-reporter/2/1_namespaces.log
_out/artifacts/k8s-reporter/2/pods
_out/artifacts/k8s-reporter/2/pods/1_kube-system_kube-proxy-d564w-kube-proxy_previous.log
[...]
```

The `partial.junit.functest.*.xml` files correspond to a ginkgo-execution node. The same applies for `_out/artifacts/k8s-reporter/*/` folders.

Finally a `junit.functest.xml` file will be created by a junit-merger utility. This final file will be recognized by deck and displayed. This is necessary to avoid having duplicate `skipped` entries, since we execute the test suite two times.

The following tests are not supposed to run in parallel:

 * Tests which change global kubevirt setting.
 * Tests wich modify kubernetes infrastructure or modify KubeVirt components.
 * Tests which use PVCs (right now, we can relax that definitely).
 * Tests which use Fedora (due to the high memory usage) or which require for other reasons more than 64Mi memory.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

Note that the first six commits are implementing the general test-setup changes. The following commits are mostly just alloing tests to run in parallel and fix minor issues on these tests if they could not be run in parallel immediately.

Example jobs which were run in paralle:

https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4051/pull-kubevirt-e2e-k8s-1.16/1305532020494962688
https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4051/pull-kubevirt-e2e-k8s-1.17/1305530259013111808
https://prow.apps.ovirt.org/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/4051/pull-kubevirt-e2e-k8s-1.18/1305604497023700992

Depending on the lane the parallel run saves between 40 and 60 minutes right now.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: hack/functests.sh
Comment (dhiller) [inline hack/functests.sh:42]: This looks weird somehow. Why is it required to pass the test binary now? Does the ginkgo test runner itself take care of the parallel execution somehow?

Comment (rmohr) [reply hack/functests.sh:42]: Directly executing the test binaries allows running the tests in non-parallel fashion only. The ginkgo binary contains the actual parallel test runner. This runner starts the test binary n-times and distributes via streams the specs to run to the binaries. I think it makes sense to also call the non-parallel execution now via the ginkgo binary to have a common way of invoking the tests.

Comment (dhiller) [reply hack/functests.sh:42]: Thanks for the explanation!

**Comment by rmohr** [comment]
/cc @ILpinto 

An early heads up, FYI.

**Comment by rmohr** [comment]
The PR would be ready for a first review. Note that the first 6 PRs implement the most important logic. The rest mostly just allows selected tests to be executed in parallel and fixe minor things on them if making them run in parallel was trivial.

**Review by davidvossel — COMMENTED** [review]
I think the PR is great. My primary concerns here are about transitioning to a model where some tests can run in parallel without impacting the stability of our test suite. 

What can we do to reduce risk here? For example would it make sense for us to only enable parallel for a single lane to start out with, and to make that lane optional?

Also, I found it surprising that we have to opt into Serial mode rather than have test cases opt into being able to run in Parallel. I wish it was the other way around so we could clearly see what tests are being opted into running in Parallel. This would also prevent any surprises for people developing new tests who wouldn't expect their tests to run in parallel by default.

File: hack/functests.sh
Comment (dhiller) [inline hack/functests.sh:69]: Maybe we should make this into an env var `KUBEVIRT_E2E_PARALLEL_NODES` so we can adjust from the prow job config?

**Review by dhiller — COMMENTED** [review]
Looks good already, just some questions.

**Comment by davidvossel** [comment]
/approve

we'll have to educate the team on how this works in order not to introduce future flakiness

File: hack/functests.sh
Comment (rmohr) [reply hack/functests.sh:49]: A debug left-over. I was mostly interested if things resolve correct.

**Review by dhiller — APPROVED** [review]
/lgtm

**Comment by rmohr** [comment]
/hold

Just holding it a little to give people on the community meeting and on kubevirt-dev time to respond to that change.

**Comment by dhiller** [comment]
One more thing: do we want to include in this PR also adding some docs to the [Testing section in getting started](https://github.com/kubevirt/kubevirt/blob/master/docs/getting-started.md#testing)?

**Comment by rmohr** [comment]
> One more thing: do we want to include in this PR also adding some docs to the [Testing section in getting started](https://github.com/kubevirt/kubevirt/blob/master/docs/getting-started.md#testing)?

Yes definitely :+1:

**Comment by dhiller** [comment]
One other thing I noticed: the test numbers of the test results you pointed to look strange (i.e. double the normal number). How is that?

**Comment by rmohr** [comment]
> One other thing I noticed: the test numbers of the test results you pointed to look strange (i.e. double the normal number). How is that?

Yes, that is due to the multiple generated junit.xml files. I can look into a way to merge them, or not print skipped tests which were run on the other test run ...

**Comment by rmohr** [comment]
> One other thing I noticed: the test numbers of the test results you pointed to look strange (i.e. double the normal number). How is that?

Fixed. Introduce a junit-merger.

**Comment by rmohr** [comment]
> One more thing: do we want to include in this PR also adding some docs to the [Testing section in getting started](https://github.com/kubevirt/kubevirt/blob/master/docs/getting-started.md#testing)?

Added it to tests/README.md

File: tools/junit-merger/junit-merger.go
Comment (dhiller) [inline tools/junit-merger/junit-merger.go:25]: This is superfluous.

File: tools/junit-merger/junit-merger.go
Comment (dhiller) [inline tools/junit-merger/junit-merger.go:80]:
```suggestion
		for i, testcase := range suite.TestCases {
```

**Review by dhiller — COMMENTED** [review]
Found two nits in the merge.

File: tools/junit-merger/junit-merger.go
Comment (rmohr) [reply tools/junit-merger/junit-merger.go:25]: Done.

File: tools/junit-merger/junit-merger.go
Comment (rmohr) [reply tools/junit-merger/junit-merger.go:80]: Done.

**Comment by dhiller** [comment]
/lgtm
/hold

Maybe this is worth a release note? Otherwise feel free to cancel the hold whenever ready.

File: tests/vmi_configuration_test.go
Comment (EdDev) [inline tests/vmi_configuration_test.go:238]: How is this change related here? (Same on the next line)

File: tests/vmi_lifecycle_test.go
Comment (EdDev) [inline tests/vmi_lifecycle_test.go:350]: Changes to this test file also seem to be unrelated in this commit.

File: tests/vm_test.go
Comment (EdDev) [inline tests/vm_test.go:982]: What cause this one to surface only now?
How will we know if this is not skipped unintentionally?

File: tests/imageupload_test.go
Comment (EdDev) [inline tests/imageupload_test.go:79]: Regarding the "Reference the default namespace directly " commit: Accessing `tests.NamespaceTestDefault` seems unsafe as it may change now. It may be safer to protect it through a function getter that validates it is "safe".
I guess in any case, we should not allow it to be referenced in a pre-populated state.

File: tests/libvmi/vmi.go
Comment (EdDev) [inline tests/libvmi/vmi.go:36]: Creating a VMI requires a namespace explicitly. How can a caller to this differently?

The problem is with the original default which is no longer valid (in parallel run), not with the definition on the VMI namespace.

**Comment by EdDev** [comment]
Running tests in parallel is a great idea, but I am a bit concern of the implications.
Unlike other parallel runs, this one shares the same target with an optimization of running each concurrent run in a different namespace.

Here are the point I see risky:
- Setting parallel run by default may cause unpredictable results when a test is added without considering this. It may collide with random runs in parallel and it may occur much later in the run. Hopefully the protection the namespace provides is enough.
- As the namespaces are now calculated during setup, accessing the variables directly is unsafe as they may change when parallel run occurs. There is a need to protect that and make sure it is not used improperly.
- Logs: How are the logs separated now? What happens with the shared system wide logs? (e.g. the journal)
- I'm guessing the CPU and Disk IO are taking a hit when running in parallel. How is that affecting the number of lane runs per HW?
- Are the serial tests ran before or after the parallel ones? How do we make sure each are starting on a clean cluster (e.g. prev cleanup succeeded/finished)?

As there is still a lot of instability at the moment, and we are going to challenge the target with new scenarios, it will be safer to try this our on dedicated lanes over an aggressive move like this. If things will not play as expected, reverting back will be painful.
At the minimum, we should not have the ipv6-kind lane around (when this gets in) which was mentioned as causing most of the problems at the moment.

File: tests/libvmi/vmi.go
Comment (rmohr) [reply tests/libvmi/vmi.go:36]: It is contraproductive if it is set in the yaml. Especially if called like this: `return New(NamespaceTestDefault, RandName(DefaultVmiName), opts...)` inside `NewFedora(opts ...Option)`.

If I would create a VM like this:

```
vm := NewFedora()
client.VirtualMachine(tests.NamespaceTestAlternative).Create(vm)
```

the create call would fail, because the namespace in the yaml is pointing to `NamespaceTestDefault` which is different to `NamespaceTestAlternative` and I would now have to do this to make it work:

```
vm := NewFedora()
vm.Spec.Template.Metadata.Namespace = tests.NamespaceTestAlternative
client.VirtualMachine(tests.NamespaceTestAlternative).Create(vm)
```

File: BUILD.bazel
Comment (rmohr) [reply BUILD.bazel:359]: Taking from `go.mod`:

```
github.com/onsi/ginkgo v1.12.1
```

which is eactly where we have our tooling library from. If we have issues with it or want to update it, it can be done in `go.mod` like before.

File: tests/vm_test.go
Comment (rmohr) [reply tests/vm_test.go:982]: Just run the tests locally on your machine with only one node and you will see it fail. In CI we always run with two nodes. It has the same risk like all the other tests which are skipped like this for the same reason.

File: tests/libvmi/vmi.go
Comment (EdDev) [reply tests/libvmi/vmi.go:36]: I am saying that the problem is in the `NewFedora` factory and not below it.
The factory should get the namespace as input explicitly and it is up to the caller to pass in the correct thing.

The current state took a shortcut as the usage did not require anything except the default.
If now it is not good enough (I guess with the computed namespace name in this case), the caller should set it explicitly.

File: tests/vmi_configuration_test.go
Comment (rmohr) [reply tests/vmi_configuration_test.go:238]: It reduces the memory footprint of the vmi configuration tests where not so much memory is needed, so that it can be run in parallel.

File: tests/libvmi/vmi.go
Comment (rmohr) [reply tests/libvmi/vmi.go:36]: > I am saying that the problem is in the `NewFedora` factory and not below it.
> The factory should get the namespace as input explicitly and it is up to the caller to pass in the correct thing.
> 
> The current state took a shortcut as the usage did not require anything except the default.
> If now it is not good enough (I guess with the computed namespace name in this case), the caller should set it explicitly.
> 
> EDIT: What are the implications of not setting the namespace in the spec?

That is the whole point. You are not supposed to set the namespace inside the yaml.

File: tests/vmi_lifecycle_test.go
Comment (rmohr) [reply tests/vmi_lifecycle_test.go:350]: The test failed on parallel runs without this. The `findLauncherForVMI` was not suitable for parallel runs. I could have fixed the function, but I decided to use the `tests.GetPodByVirtualMachineInstance` function which  we already have. Since this function expected that the pod for the VMI already exists (which `findLauncherForVMI` did not expecte), I also started to use `tests.RunVMIAndExpectScheduling`.

File: tests/imageupload_test.go
Comment (rmohr) [reply tests/imageupload_test.go:79]: > I guess in any case, we should not allow it to be referenced in a pre-populated state.

Yes. Further the parallelization model of ginkgo goes like this: It starts one binary per ginkgo node. So at least from that side there is no special protection necessary.

**Comment by rmohr** [comment]
> * Setting parallel run by default may cause unpredictable results when a test is added without considering this. It may collide with random runs in parallel and it may occur much later in the run. Hopefully the protection the namespace provides is enough.
> * As the namespaces are now calculated during setup, accessing the variables directly is unsafe as they may change when parallel run occurs. There is a need to protect that and make sure it is not used improperly.

They are set once per binary. It does not change in that sense.

> * Logs: How are the logs separated now? What happens with the shared system wide logs? (e.g. the journal)

The logs are structured like described in the PR description. node-logs will contain overlapping information but are duped per ginkgo node. 

> * I'm guessing the CPU and Disk IO are taking a hit when running in parallel. How is that affecting the number of lane runs per HW?

This almost exclusively changes CPU and memory requirements. We have spare CPU and memory. CPU is right now at roughly 50% when a node is on full capacity and the test-clusters have more than enough memory assigned already. Since we don't have memory ballooning the clusters will take all the memory which we give them at the qemu level over time. We can as well just use it for something.

The main issue is Disk IO which prevents us from spawning more clusters on the nodes. Since the serial and parallel tests use the  same cluster, this should work.

> * Are the serial tests ran before or after the parallel ones? How do we make sure each are starting on a clean cluster (e.g. prev cleanup succeeded/finished)?

As can be seen in hack/functest.sh, first the parallel ones are done (the test suite cleans up itself, as usual), then the serial ones are started. Here I slightly diverge from my previous statement that I want all tests executed. Since we can now already execute in this early state 250 tests in parallel (pretty much half of the tests), I am fine with failing early after the serial tests. We will see how this goes, but I am at least slightly optimistic that for development and flakefinder CI will still provide enough output to be valuable. I am not 100% sure that this is really the best option. May be changed again later.

> As there is still a lot of instability at the moment, and we are going to challenge the target with new scenarios, it will be safer to try this our on dedicated lanes over an aggressive move like this. If things will not play as expected, reverting back will be painful.

It is not that much of an issue to go back. Going back just means setting the environment variable which enables parallel execution to `false` for all or some lanes and fast-merge a PR outside of prow (not nice, but not hard to do).

> At the minimum, we should not have the ipv6-kind lane around (when this gets in) which was mentioned as causing most of the problems at the moment.

We will see if that is an issue. The main issue was less what we do inside the qemu-clusters, it was more about issues outside of it on the node which are created by kind. Let's see. As said, I can easily disable parallel runs in parallel for all lanes or just some.

At the end, if something in the CI environment causes issues, there is nothing wrong with the introduced changes.

File: tests/vm_test.go
Comment (rmohr) [reply tests/vm_test.go:982]: Here a WIP PR, based on this one which can eventually address these things to a degree: https://github.com/kubevirt/kubevirt/pull/4201.

File: tests/libvmi/vmi.go
Comment (EdDev) [reply tests/libvmi/vmi.go:36]: > That is the whole point. You are not supposed to set the namespace inside the yaml in this scenario. Only if you want to specify that a specific namespace needs to be used for creation.

- The VMI `New` should allow a caller to specify a specific namespace unless it is wrong to do it in general. The factory may hide it and user an empty string while others may decide to leave it up to the caller to be explicit.
- When we call `client.VirtualMachine("mynamespace").Create(vm)`, does the ns specified as the argument of `VirtualMachine` is going to overwrite what is specified in the vm spec? I do not understand how this is supposed to work. Obviously, if that argument value is enough, there is no point to set anything in the spec. Could you please clarify?

Comment (rmohr) [reply tests/libvmi/vmi.go:36]: If you set a namespace in the spec, the create calls will check if this namespace is the same which you used in the client. If they do not match the `Create` will fail. Normally one does not set the namespace in the metadata. It will be auto-filled by the cluster.

Comment (EdDev) [reply tests/libvmi/vmi.go:36]: Thank you for the clarification.

File: tests/vm_test.go
Comment (EdDev) [reply tests/vm_test.go:982]: I always run locally with two nodes, so I guess that is why I never saw this.
The pre-flight checks should do the job I guess, especially in a conformance test.

File: tests/vmi_configuration_test.go
Comment (EdDev) [reply tests/vmi_configuration_test.go:238]: Ok, please consider to add this info to the commit. When walking the code, I use the history to answer the "why" part when possible and having this info on the commit may help.

File: tests/vmi_lifecycle_test.go
Comment (EdDev) [reply tests/vmi_lifecycle_test.go:350]: Thank you for the clarification.

**Review by EdDev — COMMENTED** [review]
> > * As the namespaces are now calculated during setup, accessing the variables directly is unsafe as they may change when parallel run occurs. There is a need to protect that and make sure it is not used improperly.
> 
> They are set once per binary. It does not change in that sense.

Yes, I meant that they should not be referenced in the container bodied or definitions, only in the test running functions (BeforeEach/AfterEach/It).
Maybe initializing a structure with getters instead of simple variable may do the trick.
 
> The logs are structured like described in the PR description. node-logs will contain overlapping information but are duped per ginkgo node.

Thanks.
I think this is not only the node-logs which will contain overlapping info, but also the system and kubevirt pods which server all VMI/s / services. We will have a big challenge to debug those.

> As can be seen in hack/functest.sh, first the parallel ones are done (the test suite cleans up itself, as usual), then the serial ones are started. Here I slightly diverge from my previous statement that I want all tests executed. Since we can now already execute in this early state 250 tests in parallel (pretty much half of the tests), I am fine with failing early after the serial tests. We will see how this goes, but I am at least slightly optimistic that for development and flakefinder CI will still provide enough output to be valuable. I am not 100% sure that this is really the best option. May be changed again later.

I think we did not have to deal with post test run cleanup leftovers until now. Or did we?
Just wandering if we may fall into cases where the previous parallel run left some leftovers and did not wait for them to cleanup before continuing.

> At the end, if something in the CI environment causes issues, there is nothing wrong with the introduced changes.

The introduced changes just bring challenges to the debugging and troubleshooting steps. It also introduces new awareness need from contributors and reviewers on how to deal with it.
e2e tests are flaky because they require interaction between components, which are time/delay sensitive. When adding concurrency to the matrix, things get one level more complex and potentially flakier.

I would have favor running this for a month or so on a specific lane/staging mode and examine the results compare to the other lanes. But I can understand the more aggressive approach which can check-it-all in one shot instead.

File: tests/utils.go
Comment (rmohr) [reply tests/utils.go]: It is a convenience variable for some cleanup code. You can look for usage in utils.go.

File: tests/flags/flags.go
Comment (rmohr) [reply tests/flags/flags.go:70]: > Yes, I meant that they should not be referenced in the container bodied or definitions, only in the test running functions (BeforeEach/AfterEach/It).
> Maybe initializing a structure with getters instead of simple variable may do the trick.

I want to address that separately. We have this issue with all values which we influence via flags. I think it is enough to move all global variables to a common global config struct. A global `var TestConfig *TestConfig` should then be good enough. It would be set to the config at some point and anything which calls it before being initialized will just panic due to a nil exception.

File: hack/functests.sh
Comment (rmohr) [reply hack/functests.sh]: > > As can be seen in hack/functest.sh, first the parallel ones are done (the test suite cleans up itself, as usual), then the serial ones are started. Here I slightly diverge from my previous statement that I want all tests executed. Since we can now already execute in this early state 250 tests in parallel (pretty much half of the tests), I am fine with failing early after the serial tests. We will see how this goes, but I am at least slightly optimistic that for development and flakefinder CI will still provide enough output to be valuable. I am not 100% sure that this is really the best option. May be changed again later.
> 
> I think we did not have to deal with post test run cleanup leftovers until now. Or did we?

Not in CI on merged code. Only in the sense that we would see issues a lot on `make cluster-sync` or when running `make functest` multiple times in a row, which I do a lot locally (and others probably too). And so far the CI runs did not show anything like that.

> Just wandering if we may fall into cases where the previous parallel run left some leftovers and did not wait for them to cleanup before continuing.

The test suite creates almost not resources outside of the test namespaces. If it succeeded, it should work pretty much 100% at the moment. Nothing indicates anything different so far. That can of course change with every PR.

> 
> > At the end, if something in the CI environment causes issues, there is nothing wrong with the introduced changes.
> 
> The introduced changes just bring challenges to the debugging and troubleshooting steps. It also introduces new awareness need from contributors and reviewers on how to deal with it.
> e2e tests are flaky because they require interaction between components, which are time/delay sensitive. When adding concurrency to the matrix, things get one level more complex and potentially flakier.

Yes I took only simple tests so far. All CI runs so far did not indicate any additional flakyness (also not less). What I meant above is, even if we have to disable the parallel test runs due to CI issues, there is nothing as such wrong with this PR.

Running the tests in parallel was always a goal and k8s and others do it successfully too. There is no reason to believe that the complexity regarding to debugging for our tests is so much higher (at least not in a non-fixable way). Creating more clusters in parallel is another angle on this, we can/should leverage that too for some cases, but this is one important step in general. One can run without almost any additional resource consumption run half of the kubevirt e2e tests in less than 30 minutes. Spinning up a second cluster has huge overhead compared to that. I would still consider it, for instance to run tests which need to run in serial in another cluster in parallel.

File: tests/vmi_configuration_test.go
Comment (rmohr) [reply tests/vmi_configuration_test.go:238]: Added.

File: tests/flags/flags.go
Comment (rmohr) [reply tests/flags/flags.go:70]: Added to the commit message.

File: tests/README.md
Comment (maiqueb) [inline tests/README.md:25]: What is a destructive test ?

Comment (rmohr) [reply tests/README.md]: Should have been `[Disruptive]`. Fixed.

**Review by dhiller — APPROVED** [review]
/lgtm

**Pull request merged**
